### PR TITLE
refactor: playerList refresh

### PR DESCRIPTION
**Pull request by DieYiSu**

### CODE REVIEW

File: extensions/playerlist/common/client.mjs [hunk 1]
```diff
@@ -28,49 +28,89 @@ class Client {
...
+        playerID : player.playerID, 
+        displayName : player.displayName, 
+        displayChar : player.displayChar
+      }
+    });
+  }
+
+  /**
+   * Get the PlayerSyncMessage from server and update the playerList object.
+   * @param {object} msg - The PlayerSyncMessage from server.
```
Comment (ktpss95112) [inline extensions/playerlist/common/client.mjs:59]: `@param {PlayerSyncMessage} msg - ...`

File: extensions/playerlist/common/client.mjs [hunk 2]
```diff
@@ -28,49 +28,89 @@ class Client {
    */
   constructor(helper) {
     this.helper = helper;
+    this.template = document.getElementById('playerlist-player').cloneNode(true);
+    this.playerList = {};
```
Comment (ktpss95112) [inline extensions/playerlist/common/client.mjs:32]: Consider using `new Map()` for better performance.

File: extensions/playerlist/common/client.mjs [hunk 3]
```diff
@@ -28,49 +28,89 @@ class Client {
...
-    this.showPlayerList();
+    if (Object.keys(this.playerList).includes(msg.playerID) == false) {
+      this.playerList[msg.playerID] = {
+        playerID : msg.playerID, 
+        displayName : msg.displayName, 
+        displayChar : msg.displayChar
+      }
+    } else {
+      return;
+    }
```
Comment (ktpss95112) [inline extensions/playerlist/common/client.mjs:77]: If the type of `this.playerList` is `Map`:
```javascript
if (this.playerList.has(msg.playerID)) {
  this.playerList.set({
    ...
  });
}
// `else {return;}` is redundant
```

File: extensions/playerlist/common/client.mjs [hunk 4]
```diff
@@ -28,49 +28,89 @@ class Client {
...
-    const template = document.getElementById('playerlist-player').cloneNode(true);
-    // TODO(zeze-zeze): Optimize the way to update playerlist
+  updatePlayerList() {
     setInterval(() => {
       document.getElementById('playerlist').innerHTML = '';
-      this.helper.gameState.getPlayers().forEach(playerID => {
-        if (typeof playerID.displayName !== 'string' || !playerID.displayName.includes(document.getElementById("searchPlayer").value)) {
-          return;
+      for (const playerID of Object.keys(this.playerList)) {
+        if (typeof this.playerList[playerID].displayName !== 'string' || !this.playerList[playerID].displayName.includes(document.getElementById("searchPlayer").value)) {
```
Comment (ktpss95112) [inline extensions/playerlist/common/client.mjs:90]: If the type of `this.playerList` is `Object`:
```javascript
for (const [playerID, player] of Object.entries(this.playerList)) {
  if (typeof player.displayName !== 'string' || player.displayName.includes(this.searchPlayerDOM.value)) {
```

If the type of `this.playerList` is `Map`:
```javascript
for (const [playerID, player] of this.playerList.entries()) {
  if (typeof player.displayName !== 'string' || player.displayName.includes(this.searchPlayerDOM.value)) {
```

**Comment by john0312** [comment]
Superceded by PR #271

**Pull request closed without merging**